// vitepress-plugin-artalk/dist/index.mjs
// src/index.ts
import { stringify } from "javascript-stringify";

// src/util.ts
import path from "node:path";
import { fileURLToPath } from "node:url";
function isESM() {
  return typeof __filename === "undefined" || typeof __dirname === "undefined";
}
function getDirname() {
  return isESM() ? path.dirname(fileURLToPath(import.meta.url)) : __dirname;
}

// src/index.ts
var componentName = "ArtalkComment";
var componentFile = `${componentName}.vue`;
var aliasComponentFile = `${getDirname()}/components/${componentFile}`;
var virtualModuleId = "virtual:artalk-plugin-options";
var resolvedVirtualModuleId = `\0${virtualModuleId}`;
var slots = ["doc-after"];
function artalkPlugin(options) {
  const componentOptions = {
    ...options
  };
  let resolveConfig;
  let vitepressConfig;
  const pluginOps = {
    name: "vitepress-plugin-artalk",
    enforce: "pre",
    configResolved(config) {
      if (resolveConfig) {
        return;
      }
      resolveConfig = config;
      vitepressConfig = config.vitepress;
      if (!vitepressConfig) {
        return;
      }
      const selfTransformPageData = vitepressConfig.transformPageData;
      vitepressConfig.transformPageData = async (pageData, ctx) => {
        pageData.frontmatter.head ??= [];
        pageData.frontmatter.head.push(...getArtalkScriptHead(options));
        return selfTransformPageData?.(pageData, ctx);
      };
    },
    config: () => {
      return {
        resolve: {
          alias: {
            [`./${componentFile}`]: aliasComponentFile
          }
        }
      };
    },
    transform(code, id) {
      if (id.endsWith("vitepress/dist/client/theme-default/Layout.vue")) {
        let transformResult = code;
        for (const element of slots) {
          const slotPosition = `<slot name="${element}" />`;
          transformResult = transformResult.replace(slotPosition, `${slotPosition}<ClientOnly><${componentName} /></ClientOnly>`);
        }
        const setupPosition = '<script setup lang="ts">';
        transformResult = transformResult.replace(setupPosition, `${setupPosition}
import ${componentName} from './${componentName}.vue'`);
        return transformResult;
      }
    },
    resolveId(id) {
      if (id === virtualModuleId) {
        return resolvedVirtualModuleId;
      }
    },
    load(id) {
      if (id === resolvedVirtualModuleId) {
        return `export default ${stringify(componentOptions)}`;
      }
    }
  };
  return pluginOps;
}
function getArtalkScriptHead(options) {
  if (!options?.server) {
    return [];
  }
  const { server } = options;
  return [
    ["link", { href: `${server}/dist/Artalk.css`, rel: "stylesheet" }],
    ["script", { src: `${server}/dist/Artalk.js`, id: "artalk-script" }]
  ];
}
export {
  artalkPlugin
};
