// @sugarat/theme/node.mjs
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined")
    return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// src/utils/node/mdPlugins.ts
import { createRequire } from "module";

// ../../node_modules/.pnpm/vitepress-plugin-tabs@0.7.3_vitepress@2.0.0-alpha.16_@types+node@24.5.2_async-validator@4.2.5_wszfouv2xxx6bizfwwcmbwut3e/node_modules/vitepress-plugin-tabs/dist/node/index.js
function container_plugin(md, name, options) {
  function validateDefault(params) {
    return params.trim().split(" ", 2)[0] === name;
  }
  function renderDefault(tokens, idx, _options, env, slf) {
    if (tokens[idx].nesting === 1)
      tokens[idx].attrJoin("class", name);
    return slf.renderToken(tokens, idx, _options, env, slf);
  }
  options = options || {};
  const min_markers = 3;
  const marker_str = options.marker || ":";
  const marker_char = marker_str.charCodeAt(0);
  const marker_len = marker_str.length;
  const validate = options.validate || validateDefault;
  const render = options.render || renderDefault;
  function container(state, startLine, endLine, silent) {
    let pos;
    let auto_closed = false;
    let start = state.bMarks[startLine] + state.tShift[startLine];
    let max = state.eMarks[startLine];
    if (marker_char !== state.src.charCodeAt(start))
      return false;
    for (pos = start + 1; pos <= max; pos++)
      if (marker_str[(pos - start) % marker_len] !== state.src[pos])
        break;
    const marker_count = Math.floor((pos - start) / marker_len);
    if (marker_count < min_markers)
      return false;
    pos -= (pos - start) % marker_len;
    const markup = state.src.slice(start, pos);
    const params = state.src.slice(pos, max);
    if (!validate(params, markup))
      return false;
    if (silent)
      return true;
    let nextLine = startLine;
    for (; ; ) {
      nextLine++;
      if (nextLine >= endLine)
        break;
      start = state.bMarks[nextLine] + state.tShift[nextLine];
      max = state.eMarks[nextLine];
      if (start < max && state.sCount[nextLine] < state.blkIndent)
        break;
      if (marker_char !== state.src.charCodeAt(start))
        continue;
      if (state.sCount[nextLine] - state.blkIndent >= 4)
        continue;
      for (pos = start + 1; pos <= max; pos++)
        if (marker_str[(pos - start) % marker_len] !== state.src[pos])
          break;
      if (Math.floor((pos - start) / marker_len) < marker_count)
        continue;
      pos -= (pos - start) % marker_len;
      pos = state.skipSpaces(pos);
      if (pos < max)
        continue;
      auto_closed = true;
      break;
    }
    const old_parent = state.parentType;
    const old_line_max = state.lineMax;
    state.parentType = "container";
    state.lineMax = nextLine;
    const token_o = state.push("container_" + name + "_open", "div", 1);
    token_o.markup = markup;
    token_o.block = true;
    token_o.info = params;
    token_o.map = [startLine, nextLine];
    state.md.block.tokenize(state, startLine + 1, nextLine);
    const token_c = state.push("container_" + name + "_close", "div", -1);
    token_c.markup = state.src.slice(start, pos);
    token_c.block = true;
    state.parentType = old_parent;
    state.lineMax = old_line_max;
    state.line = nextLine + (auto_closed ? 1 : 0);
    return true;
  }
  md.block.ruler.before("fence", "container_" + name, container, { alt: [
    "paragraph",
    "reference",
    "blockquote",
    "list"
  ] });
  md.renderer.rules["container_" + name + "_open"] = render;
  md.renderer.rules["container_" + name + "_close"] = render;
}
var tabMarkerCode = "=".charCodeAt(0);
var minTabMarkerLen = 2;
var ruleBlockTab = (state, startLine, endLine, silent) => {
  let pos = state.bMarks[startLine] + state.tShift[startLine];
  const max = state.eMarks[startLine];
  if (state.parentType !== "container")
    return false;
  if (pos + minTabMarkerLen > max)
    return false;
  const marker = state.src.charCodeAt(pos);
  if (marker !== tabMarkerCode)
    return false;
  const mem = pos;
  pos = state.skipChars(pos + 1, marker);
  const tabMarkerLen = pos - mem;
  if (tabMarkerLen < minTabMarkerLen - 1)
    return false;
  if (silent)
    return true;
  let nextLine = startLine;
  let endStart = mem;
  let endPos = pos;
  for (; ; ) {
    nextLine++;
    if (nextLine >= endLine)
      break;
    endStart = state.bMarks[nextLine] + state.tShift[nextLine];
    const max$1 = state.eMarks[nextLine];
    if (endStart < max$1 && state.sCount[nextLine] < state.blkIndent)
      break;
    if (state.src.charCodeAt(endStart) !== tabMarkerCode)
      continue;
    const p = state.skipChars(endStart + 1, marker);
    if (p - endStart !== tabMarkerLen)
      continue;
    endPos = p;
    break;
  }
  const oldParent = state.parentType;
  const oldLineMax = state.lineMax;
  state.parentType = "tab";
  state.lineMax = nextLine;
  const startToken = state.push("tab_open", "div", 1);
  startToken.markup = state.src.slice(mem, pos);
  startToken.block = true;
  startToken.info = state.src.slice(pos, max).trimStart();
  startToken.map = [startLine, nextLine - 1];
  state.md.block.tokenize(state, startLine + 1, nextLine);
  const endToken = state.push("tab_close", "div", -1);
  endToken.markup = state.src.slice(endStart, endPos);
  endToken.block = true;
  state.parentType = oldParent;
  state.lineMax = oldLineMax;
  state.line = nextLine;
  return true;
};
var parseTabsParams = (input) => {
  return { shareStateKey: input.match(/key:(\S+)/)?.[1] };
};
var tabsPlugin = (md) => {
  md.use(container_plugin, "tabs", { render(tokens, index) {
    const token = tokens[index];
    if (token.nesting === 1) {
      const params = parseTabsParams(token.info);
      return `<PluginTabs ${params.shareStateKey ? `sharedStateKey="${md.utils.escapeHtml(params.shareStateKey)}"` : ""}>
`;
    } else
      return `</PluginTabs>
`;
  } });
  md.block.ruler.after("container_tabs", "tab", ruleBlockTab);
  const renderTab = (tokens, index) => {
    const token = tokens[index];
    if (token.nesting === 1) {
      const label = token.info;
      return `<PluginTabsTab ${`label="${md.utils.escapeHtml(label)}"`}>
`;
    } else
      return `</PluginTabsTab>
`;
  };
  md.renderer.rules["tab_open"] = renderTab;
  md.renderer.rules["tab_close"] = renderTab;
};

// src/utils/node/mdPlugins.ts
import timeline from "vitepress-markdown-timeline";
import { groupIconMdPlugin } from "vitepress-plugin-group-icons";

// src/utils/node/index.ts
import path from "node:path";
import { joinPath, normalizePath } from "@sugarat/theme-shared";
function aliasObjectToArray(obj) {
  return Object.entries(obj).map(([find, replacement]) => ({
    find,
    replacement
  }));
}
function isBase64ImageURL(url) {
  const regex = /^data:image\/[a-z]+;base64,/;
  return regex.test(url);
}
var imageRegex = /!\[.*?\]\((.*?)\s*(".*?")?\)/;
function getFirstImagURLFromMD(content, route) {
  const url = content.match(imageRegex)?.[1];
  const isHTTPSource = url && url.startsWith("http");
  if (!url) {
    return "";
  }
  if (isHTTPSource || isBase64ImageURL(url)) {
    return url;
  }
  const paths = joinPath("/", route).split("/");
  paths.splice(paths.length - 1, 1);
  const relativePath = url.startsWith("/") ? url : path.join(paths.join("/") || "", url);
  return joinPath("/", normalizePath(relativePath));
}
function debounce(func, delay = 1e3) {
  let timeoutId;
  return (...rest) => {
    clearTimeout(timeoutId);
    timeoutId = setTimeout(() => {
      func(...rest);
    }, delay);
  };
}
function isEqual(obj1, obj2, excludeKeys = []) {
  const keys1 = Object.keys(obj1).filter((key) => !excludeKeys.includes(key));
  const keys2 = Object.keys(obj2).filter((key) => !excludeKeys.includes(key));
  if (keys1.length !== keys2.length) {
    return false;
  }
  for (const key of keys1) {
    if (!keys2.includes(key)) {
      return false;
    }
    const val1 = obj1[key];
    const val2 = obj2[key];
    const areObjects = isObject(val1) && isObject(val2);
    if (areObjects && !isEqual(val1, val2, excludeKeys) || !areObjects && val1 !== val2) {
      return false;
    }
  }
  return true;
}
function isObject(obj) {
  return obj != null && typeof obj === "object";
}

// src/utils/node/mdPlugins.ts
function _require(module) {
  return (typeof import.meta?.url !== "undefined" ? createRequire(import.meta.url) : __require)(module);
}
function getMarkdownPlugins(cfg) {
  const markdownPlugin = [];
  if (cfg?.tabs !== false) {
    markdownPlugin.push(tabsPlugin);
  }
  if (cfg?.mermaid !== false) {
    const { MermaidMarkdown } = _require("vitepress-plugin-mermaid");
    markdownPlugin.push(MermaidMarkdown);
  }
  if (cfg?.taskCheckbox !== false) {
    markdownPlugin.push(taskCheckboxPlugin(typeof cfg?.taskCheckbox === "boolean" ? {} : cfg?.taskCheckbox));
  }
  if (cfg?.timeline !== false) {
    markdownPlugin.push(timeline);
  }
  if (cfg?.groupIcon !== false) {
    markdownPlugin.push(groupIconMdPlugin);
  }
  return markdownPlugin;
}
function taskCheckboxPlugin(ops) {
  return (md) => {
    md.use(_require("markdown-it-task-checkbox"), ops);
  };
}
function registerMdPlugins(vpCfg, plugins) {
  if (plugins.length) {
    vpCfg.markdown = {
      config(...rest) {
        plugins.forEach((plugin) => {
          plugin?.(...rest);
        });
      }
    };
  }
}
function patchMermaidPluginCfg(config) {
  if (!config.vite.resolve)
    config.vite.resolve = {};
  if (!config.vite.resolve.alias)
    config.vite.resolve.alias = {};
  config.vite.resolve.alias = [
    ...aliasObjectToArray({
      ...config.vite.resolve.alias,
      "cytoscape/dist/cytoscape.umd.js": "cytoscape/dist/cytoscape.esm.js",
      "mermaid": "mermaid/dist/mermaid.esm.mjs"
    }),
    { find: /^dayjs\/(.*).js/, replacement: "dayjs/esm/$1" }
  ];
}
function patchOptimizeDeps(config) {
  if (!config.vite.optimizeDeps) {
    config.vite.optimizeDeps = {};
  }
  config.vite.optimizeDeps.exclude = ["vitepress-plugin-tabs", "@sugarat/theme"];
}

// src/utils/node/theme.ts
import fs from "node:fs";
import { getDefaultTitle, getFileLastModifyTime, getTextSummary, getVitePressPages, grayMatter, normalizePath as normalizePath2, renderDynamicMarkdown } from "@sugarat/theme-shared";

// src/utils/client/index.ts
function shuffleArray(arr) {
  const array = [...arr];
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
}
function formatDate(d, fmt = "yyyy-MM-dd hh:mm:ss") {
  if (!(d instanceof Date)) {
    d = new Date(d);
  }
  const o = {
    "M+": d.getMonth() + 1,
    // 月份
    "d+": d.getDate(),
    // 日
    "h+": d.getHours(),
    // 小时
    "m+": d.getMinutes(),
    // 分
    "s+": d.getSeconds(),
    // 秒
    "q+": Math.floor((d.getMonth() + 3) / 3),
    // 季度
    "S": d.getMilliseconds()
    // 毫秒
  };
  if (/(y+)/.test(fmt)) {
    fmt = fmt.replace(
      RegExp.$1,
      `${d.getFullYear()}`.substr(4 - RegExp.$1.length)
    );
  }
  for (const k in o) {
    if (new RegExp(`(${k})`).test(fmt))
      fmt = fmt.replace(
        RegExp.$1,
        RegExp.$1.length === 1 ? o[k] : `00${o[k]}`.substr(`${o[k]}`.length)
      );
  }
  return fmt;
}

// src/utils/node/theme.ts
function patchDefaultThemeSideBar(cfg) {
  return cfg?.blog !== false && cfg?.recommend !== false ? {
    sidebar: [
      {
        text: "",
        items: []
      }
    ]
  } : void 0;
}
var defaultTimeZoneOffset = (/* @__PURE__ */ new Date()).getTimezoneOffset() / -60;
async function getArticleMeta(filepath, route, timeZone = defaultTimeZoneOffset, ops) {
  const fileContent = ops?.baseContent || await fs.promises.readFile(filepath, "utf-8");
  const cacheDir = ops?.cacheDir;
  const { data: frontmatter, excerpt, content } = grayMatter(fileContent, {
    excerpt: true
  });
  const meta = {
    ...frontmatter
  };
  if (!meta.title) {
    meta.title = getDefaultTitle(content);
  }
  const utcValue = timeZone >= 0 ? `+${timeZone}` : `${timeZone}`;
  const date = await (meta.date && /* @__PURE__ */ new Date(`${new Date(meta.date).toUTCString()}${utcValue}`) || getFileLastModifyTime(filepath, cacheDir));
  meta.date = formatDate(date || /* @__PURE__ */ new Date(), "yyyy/MM/dd hh:mm:ss");
  meta.categories = typeof meta.categories === "string" ? [meta.categories] : meta.categories;
  meta.tags = typeof meta.tags === "string" ? [meta.tags] : meta.tags;
  meta.tag = [meta.tag || []].flat().concat([
    .../* @__PURE__ */ new Set([...meta.categories || [], ...meta.tags || []])
  ]);
  const contentWithoutHeadings = content.replace(/^#+\s+.*$/gm, "");
  meta.description = meta.description || getTextSummary(contentWithoutHeadings, 100) || excerpt;
  meta.cover = meta.cover ?? getFirstImagURLFromMD(fileContent, route);
  if (meta.publish === false) {
    meta.hidden = true;
    meta.recommend = false;
  }
  return meta;
}
async function getArticles(cfg, vpConfig) {
  const pages = getVitePressPages(vpConfig);
  const metaResults = pages.reduce((prev, value) => {
    const { page, route, originRoute, filepath, isDynamic, dynamicRoute } = value;
    const metaPromise = isDynamic && dynamicRoute ? getArticleMeta(filepath, originRoute, cfg?.timeZone, {
      baseContent: renderDynamicMarkdown(filepath, dynamicRoute.params, dynamicRoute.content),
      cacheDir: vpConfig.cacheDir
    }) : getArticleMeta(filepath, originRoute, cfg?.timeZone, {
      cacheDir: vpConfig.cacheDir
    });
    prev[page] = {
      route,
      metaPromise
    };
    return prev;
  }, {});
  const pageData = [];
  for (const page of pages) {
    const { route, metaPromise } = metaResults[page.page];
    const meta = await metaPromise;
    if (meta.layout === "home") {
      continue;
    }
    pageData.push({
      route,
      meta
    });
  }
  return pageData;
}
function patchVPConfig(vpConfig, cfg) {
  vpConfig.head = vpConfig.head || [];
}
function patchVPThemeConfig(cfg, vpThemeConfig = {}) {
  vpThemeConfig.sidebar = patchDefaultThemeSideBar(cfg)?.sidebar;
  return vpThemeConfig;
}
function checkConfig(cfg) {
  const friendConfig = cfg?.friend;
  if (!Array.isArray(friendConfig) && friendConfig?.random) {
    friendConfig.list = shuffleArray(friendConfig.list);
  }
}

// src/utils/node/vitePlugins.ts
import {
  pagefindPlugin
} from "vitepress-plugin-pagefind";
import { RssPlugin } from "vitepress-plugin-rss";
import { cacheAllGitTimestamps, joinPath as joinPath2 } from "@sugarat/theme-shared";
import { AnnouncementPlugin } from "vitepress-plugin-announcement";
import { artalkPlugin } from "vitepress-plugin-artalk";
import { back2topPlugin } from "vitepress-plugin-back2top";
import { giscusPlugin } from "vitepress-plugin-giscus";
import { groupIconVitePlugin } from "vitepress-plugin-group-icons";
import { ImagePreviewPlugin } from "vitepress-plugin-image-preview";

// src/utils/node/hot-reload-plugin.ts
import fs2 from "fs";
import { grayMatter as grayMatter2, normalizePath as normalizePath3 } from "@sugarat/theme-shared";
function themeReloadPlugin() {
  let blogConfig;
  let vitepressConfig;
  let docsDir;
  const generateRoute = (filepath) => {
    return normalizePath3(filepath).replace(docsDir, "").replace(".md", "");
  };
  return {
    name: "@sugarat/theme-reload",
    apply: "serve",
    configureServer(server) {
      const restart = debounce(() => {
        server.restart();
      }, 500);
      server.watcher.on("add", async (path2) => {
        const route = generateRoute(path2);
        const meta = await getArticleMeta(path2, route, blogConfig?.timeZone);
        blogConfig.pagesData.push({
          route,
          meta
        });
        restart();
      });
      server.watcher.on("change", async (path2) => {
        const route = generateRoute(path2);
        const fileContent = await fs2.promises.readFile(path2, "utf-8");
        const { data: frontmatter } = grayMatter2(fileContent, {
          excerpt: true
        });
        const meta = await getArticleMeta(path2, route, blogConfig?.timeZone);
        const matched = blogConfig.pagesData.find((v) => v.route === route);
        const excludeKeys = ["date", "description"].filter((key) => !frontmatter[key]);
        const inlineKeys = [
          // vitepress 默认主题 https://vitepress.dev/zh/reference/frontmatter-config
          "lang",
          "outline",
          "head",
          "layout",
          "hero",
          "features",
          "navbar",
          "sidebar",
          "aside",
          "lastUpdated",
          "editLink",
          "footer",
          "pageClass",
          // 本主题扩展 https://theme.sugarat.top/config/frontmatter.html
          "hiddenCover",
          "readingTime",
          "buttonAfterArticle"
        ];
        if (matched && !isEqual(matched.meta, meta, inlineKeys.concat(excludeKeys))) {
          matched.meta = meta;
          restart();
        }
      });
      server.watcher.on("unlink", (path2) => {
        const route = generateRoute(path2);
        const idx = blogConfig.pagesData.findIndex((v) => v.route === route);
        if (idx >= 0) {
          blogConfig.pagesData.splice(idx, 1);
          restart();
        }
      });
    },
    configResolved(config) {
      vitepressConfig = config.vitepress;
      docsDir = vitepressConfig.srcDir;
      blogConfig = config.vitepress.site.themeConfig.blog;
    }
  };
}

// src/utils/node/vitePlugins.ts
function getVitePlugins(cfg = {}) {
  const plugins = [];
  plugins.push(cacheAllGitTimestampsPlugin());
  plugins.push(ImagePreviewPlugin(cfg?.imagePreview));
  if (cfg.themeColor) {
    plugins.push(setThemeScript(cfg.themeColor));
  }
  plugins.push(themeReloadPlugin());
  plugins.push(providePageData(cfg));
  plugins.push(coverImgTransform());
  if (cfg && cfg.search !== false) {
    const ops = cfg.search instanceof Object ? cfg.search : {};
    plugins.push(
      pagefindPlugin({
        ...ops
      })
    );
  }
  if (cfg?.mermaid !== false) {
    const { MermaidPlugin } = _require("vitepress-plugin-mermaid");
    plugins.push(inlineInjectMermaidClient());
    plugins.push(MermaidPlugin(cfg?.mermaid === true ? {} : cfg?.mermaid ?? {}));
  }
  if (cfg?.RSS) {
    ;
    [cfg?.RSS].flat().forEach((rssConfig) => plugins.push(RssPlugin(rssConfig)));
  }
  if (cfg?.comment && typeof cfg.comment !== "boolean") {
    const commentConfig = cfg.comment;
    const isArtalk = commentConfig.type === "artalk";
    if (isArtalk) {
      plugins.push(artalkPlugin(commentConfig.options));
    }
    const isGiscus = commentConfig.type === "giscus" || commentConfig.repo;
    if (isGiscus) {
      const { mobileMinify, label, icon } = commentConfig;
      const options = {
        ...commentConfig.options,
        ...commentConfig,
        mobileMinify,
        label,
        icon
      };
      plugins.push(giscusPlugin(options));
    }
  }
  if (cfg?.popover) {
    plugins.push(AnnouncementPlugin(cfg.popover));
  }
  if (cfg?.groupIcon !== false) {
    plugins.push(patchGroupIconPlugin());
    plugins.push(groupIconVitePlugin(cfg?.groupIcon));
  }
  if (cfg?.tabs !== false) {
    plugins.push(patchTabsPlugin());
  }
  if (cfg?.timeline !== false) {
    plugins.push(patchTimelinePlugin());
  }
  if (cfg?.backToTop !== false) {
    const options = typeof cfg.backToTop === "object" ? cfg.backToTop : {};
    plugins.push(back2topPlugin(options));
  }
  return plugins;
}
function patchGroupIconPlugin() {
  return createPatchPlugin({
    name: "@sugarat/theme-plugin-patch-group-icon",
    replacements: {
      "// replace-group-icon-import-code": "import 'virtual:group-icons.css'"
    }
  });
}
function patchTabsPlugin() {
  return createPatchPlugin({
    name: "@sugarat/theme-plugin-patch-tabs",
    replacements: {
      "// replace-tabs-import-code": "import { enhanceAppWithTabs } from 'vitepress-plugin-tabs/client'",
      "// replace-tabs-enhance-app-code": "enhanceAppWithTabs(ctx.app)"
    }
  });
}
function registerVitePlugins(vpCfg, plugins) {
  vpCfg.vite = {
    plugins,
    ...vpCfg.vite
  };
}
function inlineInjectMermaidClient() {
  return createPatchPlugin({
    name: "@sugarat/theme-plugin-inline-inject-mermaid-client",
    replacements: {
      "// replace-mermaid-import-code": "import Mermaid from 'vitepress-plugin-mermaid/Mermaid.vue'",
      "// replace-mermaid-mounted-code": "if (!ctx.app.component('Mermaid')) { ctx.app.component('Mermaid', Mermaid as any) }"
    }
  });
}
function createPatchPlugin({ name, replacements }) {
  return {
    name,
    enforce: "pre",
    transform(code, id) {
      if (id.endsWith("theme/src/index.ts") && code.startsWith("// @sugarat/theme index")) {
        let newCode = code;
        for (const [key, value] of Object.entries(replacements)) {
          newCode = newCode.replace(key, value);
        }
        return newCode;
      }
      return code;
    }
  };
}
function coverImgTransform() {
  let blogConfig;
  let vitepressConfig;
  let assetsDir;
  const relativeMetaName = ["cover"];
  const relativeMeta = [];
  const relativeMetaMap = {};
  const viteAssetsMap = {};
  const relativePathMap = {};
  return {
    name: "@sugarat/theme-plugin-cover-transform",
    apply: "build",
    // enforce: 'pre',
    configResolved(config) {
      vitepressConfig = config.vitepress;
      assetsDir = vitepressConfig.assetsDir;
      blogConfig = config.vitepress.site.themeConfig.blog;
      const pagesData = [...blogConfig.pagesData];
      if (vitepressConfig.site.locales && Object.keys(vitepressConfig.site.locales).length > 1 && blogConfig?.locales) {
        Object.values(blogConfig?.locales).map((v) => v.pagesData).filter((v) => !!v).forEach((v) => pagesData.push(...v));
      }
      pagesData.forEach((v) => {
        relativeMetaName.forEach((k) => {
          const value = v.meta[k];
          if (value && typeof value === "string" && value.startsWith("/")) {
            const absolutePath = `${vitepressConfig.srcDir}${value}`;
            if (relativePathMap[absolutePath]) {
              Object.assign(v.meta, {
                [k]: relativeMetaMap[absolutePath][k]
              });
            }
            relativePathMap[value] = absolutePath;
            relativePathMap[absolutePath] = value;
            if (!relativeMeta.includes(v.meta)) {
              relativeMeta.push(v.meta);
            }
            relativeMetaMap[absolutePath] = v.meta;
          }
        });
      });
    },
    moduleParsed(info) {
      if (!relativePathMap[info.id]) {
        return;
      }
      const asset = info.code?.match(/export default "(.*)"/)?.[1];
      if (!asset) {
        return;
      }
      viteAssetsMap[info.id] = asset;
      viteAssetsMap[asset] = info.id;
      relativeMeta.forEach((meta) => {
        relativeMetaName.forEach((k) => {
          const value = meta[k];
          if (!value || !relativePathMap[value]) {
            return;
          }
          const viteAsset = viteAssetsMap[relativePathMap[value]];
          if (viteAsset) {
            Object.assign(meta, { [k]: viteAsset });
          }
        });
      });
    },
    generateBundle(_, bundle) {
      const assetsMap = Object.entries(bundle).filter(([key]) => {
        return key.startsWith(assetsDir);
      }).map(([_2, value]) => {
        return value;
      }).filter((v) => v.type === "asset");
      if (!assetsMap.length) {
        return;
      }
      relativeMeta.forEach((meta) => {
        relativeMetaName.forEach((k) => {
          const value = meta[k];
          if (!value || !viteAssetsMap[value]) {
            return;
          }
          const absolutePath = viteAssetsMap[value];
          const matchAsset = assetsMap.find((v) => joinPath2(`${vitepressConfig.srcDir}/`, v.originalFileName) === absolutePath);
          if (matchAsset) {
            Object.assign(meta, { [k]: joinPath2("/", matchAsset.fileName) });
          }
        });
      });
    }
  };
}
function providePageData(cfg) {
  return {
    name: "@sugarat/theme-plugin-provide-page-data",
    async config(config, env) {
      const vitepressConfig = config.vitepress;
      const pagesData = await getArticles(cfg, vitepressConfig);
      if (vitepressConfig.site.locales && Object.keys(vitepressConfig.site.locales).length > 1) {
        if (!vitepressConfig.site.themeConfig.blog.locales) {
          vitepressConfig.site.themeConfig.blog.locales = {};
        }
        const localeKeys = Object.keys(vitepressConfig.site.locales);
        localeKeys.forEach((localeKey) => {
          if (!vitepressConfig.site.themeConfig.blog.locales[localeKey]) {
            vitepressConfig.site.themeConfig.blog.locales[localeKey] = {};
          }
          vitepressConfig.site.themeConfig.blog.locales[localeKey].pagesData = pagesData.filter((v) => {
            const { route } = v;
            const isRoot = localeKey === "root";
            if (isRoot) {
              return !localeKeys.filter((v2) => v2 !== "root").some((k) => route.startsWith(`/${k}`));
            }
            return route.startsWith(`/${localeKey}`);
          });
        });
        if (env.mode === "production") {
          return;
        }
      }
      vitepressConfig.site.themeConfig.blog.pagesData = pagesData;
    }
  };
}
function cacheAllGitTimestampsPlugin() {
  return {
    name: "@sugarat/theme-plugin-cache-all-git-timestamps",
    async config(config) {
      const { srcDir } = config.vitepress;
      await cacheAllGitTimestamps(srcDir);
    }
  };
}
function setThemeScript(themeColor) {
  let resolveConfig;
  const pluginOps = {
    name: "@sugarat/theme-plugin-theme-color-script",
    enforce: "pre",
    configResolved(config) {
      if (resolveConfig) {
        return;
      }
      resolveConfig = config;
      const vitepressConfig = config.vitepress;
      if (!vitepressConfig) {
        return;
      }
      const selfTransformHead = vitepressConfig.transformHead;
      vitepressConfig.transformHead = async (ctx) => {
        const selfHead = await Promise.resolve(selfTransformHead?.(ctx)) || [];
        return selfHead.concat([
          ["script", { type: "text/javascript" }, `;(function() {
            document.documentElement.setAttribute("theme", "${themeColor}");
          })()`]
        ]);
      };
    }
  };
  return pluginOps;
}
function patchTimelinePlugin() {
  return createPatchPlugin({
    name: "@sugarat/theme-plugin-patch-timeline",
    replacements: {
      "// replace-timeline-import-code": "import 'vitepress-markdown-timeline/dist/theme/index.css'"
    }
  });
}

// src/node.ts
function getThemeConfig(cfg = {}) {
  checkConfig(cfg);
  cfg.mermaid = cfg.mermaid ?? false;
  const pagesData = [];
  const extraVPConfig = {
    vite: {
      // see https://sass-lang.com/documentation/breaking-changes/legacy-js-api/
      // css: {
      //   preprocessorOptions: {
      //     scss: {
      //       api: 'modern',
      //     },
      //   },
      // },
      build: {
        // https://vite.dev/config/build-options.html#build-chunksizewarninglimit
        chunkSizeWarningLimit: 2048
      }
    }
  };
  const vitePlugins = getVitePlugins(cfg);
  registerVitePlugins(extraVPConfig, vitePlugins);
  const markdownPlugin = getMarkdownPlugins(cfg);
  registerMdPlugins(extraVPConfig, markdownPlugin);
  if (cfg?.mermaid !== false) {
    patchMermaidPluginCfg(extraVPConfig);
  }
  patchOptimizeDeps(extraVPConfig);
  patchVPConfig(extraVPConfig, cfg);
  return {
    themeConfig: {
      blog: {
        pagesData,
        // 插件里补全
        ...cfg
      },
      // 补充一些额外的配置用于继承
      ...patchVPThemeConfig(cfg)
    },
    ...extraVPConfig
  };
}
function defineConfig(config) {
  return config;
}
function defineLocaleConfig(cfg) {
  return cfg;
}
function footerHTML(footerData) {
  const data = [footerData || []].flat();
  return data.map((d) => {
    const { icon, text, link } = d;
    return `<span class="footer-item">
    ${icon ? `<i>${icon}</i>` : ""}
    ${link ? `<a href="${link}" target="_blank" rel="noopener noreferrer">${text}</a>` : `<span>${text}</span>`}
</span>`;
  }).join("");
}
export {
  defineConfig,
  defineLocaleConfig,
  footerHTML,
  getThemeConfig
};
