// vitepress-plugin-image-preview/dist/index.mjs
// src/index.ts
import path from "node:path";
import { fileURLToPath } from "node:url";
import { stringify } from "javascript-stringify";
function isESM() {
  return typeof __filename === "undefined" || typeof __dirname === "undefined";
}
function getDirname() {
  return isESM() ? path.dirname(fileURLToPath(import.meta.url)) : __dirname;
}
var componentName = "ImagePreview";
var componentFile = `${componentName}.vue`;
var aliasComponentFile = `${getDirname()}/components/${componentFile}`;
var virtualModuleId = "virtual:image-preview-options";
var resolvedVirtualModuleId = `\0${virtualModuleId}`;
function ImagePreviewPlugin(options = {}) {
  const pluginOps = {
    name: "vitepress-plugin-image-preview",
    enforce: "pre",
    config: () => {
      return {
        resolve: {
          alias: {
            [`./${componentFile}`]: aliasComponentFile
          }
        }
      };
    },
    transform(code, id) {
      if (id.endsWith("vitepress/dist/client/theme-default/Layout.vue")) {
        let transformResult = code;
        const slots = [options.slots || ["doc-before", "page-top"]].flat();
        for (const slot of slots) {
          const slotPosition = `<slot name="${slot}" />`;
          transformResult = transformResult.replace(slotPosition, `${slotPosition}
<ClientOnly><${componentName} /></ClientOnly>`);
        }
        const setupPosition = '<script setup lang="ts">';
        transformResult = transformResult.replace(setupPosition, `${setupPosition}
import ${componentName} from './${componentFile}'`);
        return transformResult;
      }
    },
    resolveId(id) {
      if (id === virtualModuleId) {
        return resolvedVirtualModuleId;
      }
    },
    load(id) {
      if (id === resolvedVirtualModuleId) {
        return `export default ${stringify(options)}`;
      }
    }
  };
  return pluginOps;
}
export {
  ImagePreviewPlugin
};
